src/harness/sas_test_harness.py

```diff
@@ -0,0 +1,311 @@
+    self.daemon = True
+    self.server = None
+
+
+  def getSasThUrl(self):
+    return self.http_server_url
+
+  def getUrlHostNameAndPort(self):
+    """ This method will return the http_server_url used in startServer method by truncating other
+    querystring in url and returns tuple(<hostname>, <port>)
+    """
+    splitted_url = self.http_server_url.split('/')[2:]
+    logging.info(splitted_url)
+    return splitted_url[0].split(':')[0], int(splitted_url[0].split(':')[1])
+
+  def getDumpFilePath(self):
+    """ This method will return absolute path where FAD Records of SAS-TH are stored.
+    """
+    harness_dir = os.path.dirname(os.path.abspath(inspect.getfile(inspect.currentframe())))
+    dumpfile_path = os.path.join(harness_dir, 'FADRecords', self.dump_path)
+    return dumpfile_path
+
+  def run(self):
+    self.startServer()
+
+  def shutdown(self):
+    self.stopServer()
+
+  def startServer(self):
+    """ Starts the HTTPServer as background thread.
+    """
+    SasTestHarnessServerHandler.setupDumpFilePath(self.getDumpFilePath())
+    self.server = HTTPServer(self.getUrlHostNameAndPort(), SasTestHarnessServerHandler)
+    self.server.socket = ssl.wrap_socket(
+        self.server.socket,
+        certfile=self.cert_file,
+        keyfile=self.key_file,
+        ca_certs=CA_CERT,
+        cert_reqs=ssl.CERT_REQUIRED,
+        ssl_version=ssl.PROTOCOL_TLSv1_2,
+        ciphers=':'.join(CIPHERS),
+        server_side=True
+        )
+    # Start HTTP Server
+    logging.info("Started Test Harness Server:%s at %s", self.name, self.http_server_url)
+    self.server.serve_forever()
+
+  def stopServer(self):
+    """ This method used to stop http_server  """
+    self.server.shutdown()
+    logging.info("Stopped Test Harness Server:%s", self.name)
+
+  def generateFadRecords(self, cbsd_records, grant_records_list, ppa_records,
+                         esc_sensor_records):
+    """ Generates the cbsd,ppa and esc_sensor dump file and dumps the FAD file in specified path
+
+    Args:
+       cbsd_records: list of cbsdData objects that includes one or more grants object and FAD
+       grant_records_list: list of muliple grant records for all cbsd devices
+         For an example : muliple grants [ [grant1, grant2], [grant1, grant2] ] associated with each
+         cbsd records [ cbsd1, cbsd2 ]
+       ppa_records:  list of ppa_record objects used to generate ppa_dump file and FAD
+       esc_sensor_records: list of esc_sensor_records objects used to generate
+                            esc_sensor_records and FAD
+
+    Returns:
+       It returns None and successfully generates the dumps in specified file path
+    """
+
+    def getCbsdReferenceId(fcc_id, serial_number):
+      """creates encoded cbsd_reference_id using sha1 with fcc_id and serial number of cbsd
+      """
+      return str(fcc_id + '/' + str(hashlib.sha1(serial_number).hexdigest())).encode('utf-8')
+
+    def writeDumpFile(file_path, data):
```

Comment:
This   method takes the absolute file path as input and was written to be more   generic. As this method is currently used to write files only in the base   dump directory, we can change it to accept just the filename  instead of absolute path.     This method has been changed accordingly.